File: Java/Job Scheduling/priority_scheduling.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.Scanner;

public class priority_scheduling {
    static class PriorityScheduling {
        public String name;
        public int completionTime, arrivalTime, burstTime, turnAroundTime, responseTime, waitingTime;
        public int priority;
    }

    static class PriorityComparator implements Comparator<PriorityScheduling> {
        private int curr;

        public PriorityComparator(int curr) {
            this.curr = curr;
        }

        @Override
        public int compare(PriorityScheduling o1, PriorityScheduling o2) {
            if (curr >= o2.arrivalTime && curr >= o1.arrivalTime) {
                return Integer.compare(o1.priority, o2.priority);
            }
            return 0;
        }
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        System.out.print("Enter Number of Processes: ");
        int n = scanner.nextInt();

        List<PriorityScheduling> process = new ArrayList<>();

        System.out.print("Enter Name of Processes: ");
        for (int i = 0; i < n; i++) {
            PriorityScheduling ps = new PriorityScheduling();
            ps.name = scanner.next();
            process.add(ps);
        }
        System.out.print("Enter Arrival Time of Processes: ");
        for (int i = 0; i < n; i++) {
            process.get(i).arrivalTime = scanner.nextInt();
        }
        System.out.print("Enter Priority of Processes: ");
        for (int i = 0; i < n; i++) {
            process.get(i).priority = scanner.nextInt();
        }
        System.out.print("Enter Burst Time of Processes: ");
        for (int i = 0; i < n; i++) {
            process.get(i).burstTime = scanner.nextInt();
        }

        int curr = 0;
        double avgTurnAroundTime = 0, avgWaitingTime = 0, avgResponseTime = 0;

        for (int i = 0; i < n; i++) {
            Collections.sort(process, new PriorityComparator(curr));

            process.get(i).responseTime = curr - process.get(i).arrivalTime;
            process.get(i).completionTime = curr + process.get(i).burstTime;

            curr = process.get(i).completionTime;

            process.get(i).turnAroundTime = process.get(i).completionTime - process.get(i).arrivalTime;
            process.get(i).waitingTime = process.get(i).turnAroundTime - process.get(i).burstTime;

            avgTurnAroundTime += process.get(i).turnAroundTime;
            avgWaitingTime += process.get(i).waitingTime;
            avgResponseTime += process.get(i).responseTime;
        }

        System.out.println("    AT\tP\tBT\tCT\tTAT\tWT\tRT");
        for (int i = 0; i < n; i++) {
            PriorityScheduling ps = process.get(i);
            System.out.printf("%s:  %d\t%d\t%d\t%d\t%d\t%d\t%d\n", ps.name, ps.arrivalTime, ps.priority, ps.burstTime,
                    ps.completionTime, ps.turnAroundTime, ps.waitingTime, ps.responseTime);
        }

        System.out.println();
        System.out.printf("Average TurnAroundTime: %.2f\n", avgTurnAroundTime / n);
        System.out.printf("Average WaitingTime: %.2f\n", avgWaitingTime / n);
        System.out.printf("Average ResponseTime: %.2f\n", avgResponseTime / n);

        System.out.println();
        System.out.println("Gantt Chart");
        System.out.print(process.get(0).arrivalTime + "  |" + process.get(0).name + "|  ");
        for (int i = 0; i < n - 1; i++) {
            System.out.print(process.get(i).completionTime + "  |" + process.get(i + 1).name + "|  ");
        }
        System.out.print(process.get(n - 1).completionTime);
        System.out.println();
    }
}

File: Python/Job Scheduling/priority_scheduling.py
class PriorityScheduling:
    def __init__(self):
        self.name = ""
        self.completionTime = 0
        self.arrivalTime = 0
        self.burstTime = 0
        self.turnAroundTime = 0
        self.responseTime = 0
        self.waitingTime = 0
        self.priority = 0


curr = 0

if __name__ == "__main__":
    avgTurnAroundTime = 0
    avgWaitingTime = 0
    avgResponseTime = 0

    print("Enter Number of Processes: ")
    n = int(input())

    process = []

    print("Enter Name of Processes: ")
    for i in range(n):
        p = PriorityScheduling()
        p.name = input()
        process.append(p)

    print("Enter Arrival Time of Processes: ")
    for i in range(n):
        process[i].arrivalTime = int(input())

    print("Enter Priority of Processes: ")
    for i in range(n):
        process[i].priority = int(input())

    print("Enter Burst Time of Processes: ")
    for i in range(n):
        process[i].burstTime = int(input())

    for i in range(n):
        process.sort(key=lambda x: x.priority-curr)
        
        process[i].responseTime = curr - process[i].arrivalTime
        process[i].completionTime = curr + process[i].burstTime

        curr = process[i].completionTime

        process[i].turnAroundTime = process[i].completionTime - \
            process[i].arrivalTime
        process[i].waitingTime = process[i].turnAroundTime - \
            process[i].burstTime

        avgTurnAroundTime += process[i].turnAroundTime
        avgResponseTime += process[i].responseTime
        avgWaitingTime += process[i].waitingTime

    print("    AT\tP\tBT\tCT\tTAT\tWT\tRT")
    for i in range(n):
        print(process[i].name + ":  " + str(process[i].arrivalTime) + "\t" + str(process[i].priority) + "\t" + str(process[i].burstTime) + "\t" + str(
            process[i].completionTime) + "\t" + str(process[i].turnAroundTime) + "\t" + str(process[i].waitingTime) + "\t" + str(process[i].responseTime))

    print()
    print("Average TurnAroundTime: " + str(avgTurnAroundTime / n))
    print("Average WaitingTime: " + str(avgWaitingTime / n))
    print("Average ResponseTime: " + str(avgResponseTime / n))

    print()
    print("Gantt Chart")
    print(str(process[0].arrivalTime) + "  |" +
          process[0].name + "|  ", end="")
    for i in range(n - 1):
        print(str(process[i].completionTime) + "  |" +
              process[i + 1].name + "|  ", end="")
    print(str(process[n - 1].completionTime))
